server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java

```diff
@@ -317,24 +317,29 @@ protected void serverAcceptedChannel(HttpChannel httpChannel) {
         boolean addedOnThisCall = httpChannels.add(httpChannel);
         assert addedOnThisCall : "Channel should only be added to http channel set once";
         totalChannelsAccepted.incrementAndGet();
-        httpChannelStats.put(
-            HttpStats.ClientStats.getChannelKey(httpChannel),
-            new HttpStats.ClientStats(threadPool.absoluteTimeInMillis())
-        );
-        httpChannel.addCloseListener(ActionListener.wrap(() -> {
-            try {
-                httpChannels.remove(httpChannel);
-                HttpStats.ClientStats clientStats = httpChannelStats.get(HttpStats.ClientStats.getChannelKey(httpChannel));
-                if (clientStats != null) {
-                    clientStats.closedTimeMillis = threadPool.absoluteTimeInMillis();
+        addClientStats(httpChannel);
+        logger.trace(() -> new ParameterizedMessage("Http channel accepted: {}", httpChannel));
+    }
+
+    private HttpStats.ClientStats addClientStats(final HttpChannel httpChannel) {
+        HttpStats.ClientStats clientStats = new HttpStats.ClientStats(threadPool.absoluteTimeInMillis());
+        httpChannelStats.put(HttpStats.ClientStats.getChannelKey(httpChannel), clientStats);
+        if (httpChannel != null) {
```

Comment:
Not that I expect it to be null, but these lines should be inside the `if`, or else it could end up never having a close time because the listener is not added

```suggestion
        if (httpChannel != null) {
            HttpStats.ClientStats clientStats = new HttpStats.ClientStats(threadPool.absoluteTimeInMillis());
            httpChannelStats.put(HttpStats.ClientStats.getChannelKey(httpChannel), clientStats);
```

We can have either a `final` var for `clientStats` set here and in an `else` clause, or just return `null` if the `httpChannel` is null (since the return value is already protected by checking if the channel is null